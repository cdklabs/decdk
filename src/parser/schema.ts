export namespace schema {
  export interface Template {
    readonly Resources?: Record<string, Resource>;
    readonly AWSTemplateFormatVersion?: string;
    readonly Description?: string;
    readonly Metadata?: Record<any, any>;
    readonly Parameters?: Record<string, Parameter>;
    readonly Rules?: Record<string, Rule>;
    readonly Mappings?: Record<string, Mapping>;
    readonly Conditions?: Record<string, Condition>;
    readonly Transform?: string[];
    readonly Outputs?: Record<string, Output>;
  }

  export interface Resource {
    readonly Type: string;
    readonly Properties?: Record<string, CfnValue<any>>;
    readonly Condition?: string;
    readonly DependsOn?: string | string[];
    readonly DeletionPolicy?: 'Retain' | 'Delete' | 'Snapshot';
    readonly UpdateReplacePolicy?: 'Retain' | 'Delete' | 'Snapshot';
    readonly Metadata?: Record<string, any>;
    readonly CreationPolicy?: CreationPolicy;
    readonly UpdatePolicy?: UpdatePolicy;
    readonly Tags?: Tag[];
    readonly Overrides?: Override[];
    readonly On?: string;
    readonly Call?: CfnValue<any>;
  }

  export interface CreationPolicy {
    readonly AutoScalingCreationPolicy?: AutoScalingCreationPolicy;
    readonly ResourceSignal?: ResourceSignal;
  }

  export interface UpdatePolicy {
    // @todo
    [key: string]: any;
  }

  export interface AutoScalingCreationPolicy {
    readonly MinSuccessfulInstancesPercent: number;
  }

  export interface ResourceSignal {
    readonly Count: number;
    readonly Timeout: string;
  }

  export interface Parameter {
    readonly Type: string;
    readonly Description?: string;
    readonly Default?: any;
    readonly AllowedValues?: any[];
    readonly AllowedPattern?: string;
    readonly ConstraintDescription?: string;
    readonly MaxLength?: number;
    readonly MinLength?: number;
    readonly MinValue?: number;
    readonly MaxValue?: number;
    readonly NoEcho?: boolean;
  }

  export interface Rule {}

  export interface Output {
    readonly Description?: string;
    readonly Value: CfnValue<string>;
    readonly Export?: {
      readonly Name: CfnValue<string>;
    };
    readonly Condition?: string;
  }

  export interface Tag {
    readonly Key: string;
    readonly Value: string;
  }

  export interface Override {
    readonly ChildConstructPath?: string;
    readonly Update?: {
      readonly Path: string;
      readonly Value: CfnValue<any>;
    };
    readonly Delete?: {
      readonly Path: string;
    };
    readonly RemoveResource?: boolean;
  }

  export type Mapping = Record<string, any>;

  export type CfnValue<A> = A | Intrinsic;

  export type Intrinsic =
    | Ref
    | FnBase64
    | FnCidr
    | FnFindInMap
    | FnGetAZs
    | FnGetAtt
    | FnIf
    | FnImportValue
    | FnJoin
    | FnSelect
    | FnSplit
    | FnSub;

  export type Ref = { Ref: string };
  export type FnBase64 = { 'Fn::Base64': CfnValue<string> };
  export type FnCidr = {
    'Fn::Cidr': [string | FnSelect | Ref, number, number];
  };
  export type FnFindInMap = {
    'Fn::FindInMap': [CfnValue<string>, CfnValue<string>, CfnValue<string>];
  };
  export type FnGetAZs = { 'Fn::GetAZs': string | Ref };
  export type FnGetAtt = { 'Fn::GetAtt': string | [string, string | Ref] };
  export type FnIf = { 'Fn::If': [string, CfnValue<any>, CfnValue<any>] };
  export type FnImportValue = { 'Fn::ImportValue': CfnValue<string> };
  export type FnJoin = {
    'Fn::Join': [string, CfnValue<Array<CfnValue<string>>>];
  };
  export type FnSelect = { 'Fn::Select': [CfnValue<number>, CfnValue<string>] };
  export type FnSplit = { 'Fn::Split': [string, CfnValue<string>] };
  export type FnSub = {
    'Fn::Sub': string | [string, Record<string, CfnValue<string>>];
  };

  export type FnTransform = {
    'Fn::Transform': {
      readonly Name: string;
      readonly Parameters: Record<string, CfnValue<any>>;
    };
  };

  export type Condition = FnEquals | FnAnd | FnNot | FnOr;

  export type ConditionValue<A> = A | FnFindInMap | Ref;

  export type FnAnd = { 'Fn::And': Condition[] };
  export type FnOr = { 'Fn::Or': Condition[] };
  export type FnNot = { 'Fn::Not': [Condition] };
  export type FnEquals = {
    'Fn::Equals': [ConditionValue<string>, ConditionValue<string>];
  };
}
