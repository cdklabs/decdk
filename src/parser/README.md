# Template parser

This folder contains the logic to parse declarative CDK templates. It transforms the input file into an abstract 
syntax tree, with all the necessary annotations, that can be used later for evaluation.

This code was copied and adapted from [cfngine], by Rico Huijbers.


[cfngine]: https://github.com/rix0rrr/cfngine