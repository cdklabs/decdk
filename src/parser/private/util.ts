export function mkDict<A>(
  xs: ReadonlyArray<readonly [string, A]>
): Record<string, A> {
  const ret: Record<string, A> = {};
  for (const [k, v] of xs) {
    ret[k] = v;
  }
  return ret;
}
