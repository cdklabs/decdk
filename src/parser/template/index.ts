export * from './enums';
export * from './expression';
export * from './parameters';
export * from './resource';
export * from './template';
