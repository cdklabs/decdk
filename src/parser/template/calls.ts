import {
  ArrayLiteral,
  parseExpression,
  TemplateExpression,
} from './expression';
import { assertOneField, assertString } from '../private/types';

export interface FactoryMethodCall {
  readonly target?: string;
  readonly methodName: string;
  readonly arguments: ArrayLiteral;
}

export function parseCall(x: unknown): FactoryMethodCall | undefined {
  if (x == null) return undefined;

  const array = Array.isArray(x) ? x : [x];
  switch (array.length) {
    case 1:
      return parseStaticCall(array);
    case 2:
      return parseInstanceCall(array);
    default:
      throw new SyntaxError(
        `Method calls should have 1 or 2 elements, got ${array}`
      );
  }
}

function parseStaticCall(array: any[]) {
  const methodFqn = assertOneField(array[0]);
  return {
    methodName: methodFqn,
    arguments: toArrayLiteral(parseExpression(array[0][methodFqn])),
  };
}

function parseInstanceCall(array: any[]) {
  const methodFqn = assertOneField(array[1]);
  return {
    target: assertString(array[0]),
    methodName: methodFqn,
    arguments: toArrayLiteral(parseExpression(array[1][methodFqn])),
  };
}

export function toArrayLiteral(x: TemplateExpression): ArrayLiteral {
  return x.type === 'array'
    ? x
    : {
        type: 'array',
        array: [x],
      };
}
