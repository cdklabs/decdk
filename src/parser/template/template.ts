import { promises as fs } from 'fs';
import { parseCfnYaml } from '../private/cfn-yaml';
import { DependencyGraph } from '../private/toposort';
import { schema } from '../schema';
import { parseExpression, TemplateExpression } from './expression';
import { parseMapping, TemplateMapping } from './mappings';
import { parseOutput, TemplateOutput } from './output';
import { TemplateParameters } from './parameters';
import { parseTemplateResource, TemplateResource } from './resource';

/**
 * A template describes the desired state of some infrastructure
 */
export class Template {
  public static async fromFile(fileName: string): Promise<Template> {
    const tpl = parseCfnYaml(
      await fs.readFile(fileName, { encoding: 'utf-8' })
    );
    if (!tpl.Resources) {
      throw new Error(`${fileName}: does not look like a template`);
    }
    return new Template(tpl);
  }

  public static empty(): Template {
    return new Template({
      Resources: {},
    });
  }

  public readonly parameters: TemplateParameters;
  public readonly resources: Map<string, TemplateResource>;
  public readonly conditions: Map<string, TemplateExpression>;
  public readonly mappings: Map<string, TemplateMapping>;
  public readonly outputs: Map<string, TemplateOutput>;

  constructor(private readonly template: schema.Template) {
    this.parameters = new TemplateParameters(this.template.Parameters ?? {});

    this.resources = new Map(
      Object.entries(template.Resources ?? {}).map(([k, v]) => [
        k,
        parseTemplateResource(v),
      ])
    );

    this.conditions = new Map(
      Object.entries(template.Conditions ?? {}).map(([k, v]) => [
        k,
        parseExpression(v),
      ])
    );

    this.mappings = new Map(
      Object.entries(template.Mappings ?? {}).map(([k, v]) => [
        k,
        parseMapping(v),
      ])
    );

    this.outputs = new Map(
      Object.entries(template.Outputs ?? {}).map(([k, v]) => [
        k,
        parseOutput(v),
      ])
    );
  }

  public resource(logicalId: string) {
    const r = this.resources.get(logicalId);
    if (!r) {
      console.log(this.resources);
      console.log(new Error('').stack);
      throw new Error(`No such resource: ${logicalId}`);
    }
    return r;
  }

  public condition(logicalId: string) {
    const condition = this.conditions.get(logicalId);
    if (!condition) {
      throw new Error(`No such Condition: ${logicalId}`);
    }
    return condition;
  }

  /**
   * Return a graph containing all dependencies in execution order
   *
   * This includes relations established both by property references, as
   * well as 'DependsOn' declarations.
   */
  public resourceGraph(): DependencyGraph<TemplateResource> {
    const dependencies: Map<string, any> = new Map();
    for (const [k, v] of this.resources.entries()) {
      dependencies.set(k, v.dependencies);
    }

    return new DependencyGraph(
      Object.fromEntries(this.resources.entries()),
      dependencies
    );
  }
}
