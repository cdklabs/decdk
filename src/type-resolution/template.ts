import * as reflect from 'jsii-reflect';
import {
  toTypedTemplateExpression,
  TypedTemplateExpression,
  TypedTemplateOutput,
} from './expression';
import { resolveResourceLike, ResourceLike } from './resource-like';
import { AnnotationsContext } from '../error-handling';
import { DependencyGraph } from '../parser/private/toposort';
import { Template, TemplateParameter } from '../parser/template';
import { TemplateHook } from '../parser/template/hooks';
import { TemplateMapping } from '../parser/template/mappings';
import { TemplateRule } from '../parser/template/rules';

export interface TypedTemplateProps {
  typeSystem: reflect.TypeSystem;
  annotations?: AnnotationsContext;
}

/**
 * A template describes the desired state of some infrastructure
 */
export class TypedTemplate {
  public readonly description?: string;
  public readonly templateFormatVersion?: string;
  public readonly resources: DependencyGraph<ResourceLike>;
  public readonly parameters: Map<string, TemplateParameter>;
  public readonly conditions: Map<string, TypedTemplateExpression>;
  public readonly mappings: Map<string, TemplateMapping>;
  public readonly outputs: Map<string, TypedTemplateOutput>;
  public readonly transform: string[];
  public readonly metadata: Map<string, unknown>;
  public readonly rules: Map<string, TemplateRule>;
  public readonly hooks: Map<string, TemplateHook>;

  constructor(public template: Template, props: TypedTemplateProps) {
    const { annotations = AnnotationsContext.root(), typeSystem } = props;

    const ctx = {
      annotations,
      typeSystem,
      template: this,
    };

    const resourceGraph = template.resourceGraph();
    this.resources = new DependencyGraph();
    resourceGraph.forEach((logicalId, resource) => {
      const node = resolveResourceLike(logicalId, resource, ctx);
      this.resources.setNode(
        logicalId,
        node,
        new Set(resourceGraph.dependencies.get(logicalId))
      );
      return node;
    });

    this.description = template.description;
    this.templateFormatVersion = template.templateFormatVersion;
    this.parameters = template.parameters;
    this.conditions = new Map();
    for (let [logicalId, condition] of template.conditions) {
      const typedCondition = toTypedTemplateExpression(condition);
      this.conditions.set(logicalId, typedCondition);
    }
    this.mappings = template.mappings;
    this.transform = template.transform;
    this.metadata = template.metadata;
    this.rules = template.rules;
    this.hooks = template.hooks;
    this.outputs = new Map();
    for (let [logicalId, output] of template.outputs) {
      const typedOutput = {
        exportName: output.exportName
          ? toTypedTemplateExpression(output.exportName)
          : output.exportName,
        conditionName: output.conditionName,
        value: toTypedTemplateExpression(output.value),
        description: output.description,
      } as TypedTemplateOutput;
      this.outputs.set(logicalId, typedOutput);
    }
  }

  public resource(logicalId: string): ResourceLike {
    return this.resources.get(logicalId);
  }
}
