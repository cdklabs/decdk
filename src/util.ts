import * as path from 'path';
import * as jsiiReflect from 'jsii-reflect';
import { Template } from './parser/template';

/**
 * Reads a YAML/JSON template file.
 */
export async function readTemplate(templateFile: string): Promise<Template> {
  return Template.fromFile(templateFile);
}

export async function loadTypeSystem(validate = true) {
  const typeSystem = new jsiiReflect.TypeSystem();
  await typeSystem.loadNpmDependencies(path.resolve(__dirname, '..'), {
    validate,
  });
  return typeSystem;
}

export function stackNameFromFileName(fileName: string) {
  return path.parse(fileName).name.replace('.', '-');
}
