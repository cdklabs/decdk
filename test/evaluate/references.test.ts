import { Match } from 'aws-cdk-lib/assertions';
import { expect } from 'expect';
import { Template } from '../../src/parser/template';
import { Testing } from '../util';

suite('Evaluate References', () => {
  test('can references L2 construct from L1 resource', async () => {
    // GIVEN
    const template = await Testing.template(
      Template.fromObject({
        Resources: {
          MyBucket: {
            Type: 'aws-cdk-lib.aws_s3.Bucket',
          },
          MyFunction: {
            Type: 'AWS::Lambda::Function',
            Properties: {
              Runtime: 'nodejs16.x',
              Handler: 'index.handler',
              Code: {
                ZipFile:
                  'exports.handler = function(event) { console.log("hello world!"); }',
              },
              Description: { Ref: 'MyBucket' },
            },
          },
        },
      })
    );

    // THEN
    template.hasResourceProperties('AWS::Lambda::Function', {
      Description: {
        Ref: Match.stringLikeRegexp('^MyBucket.{8}$'),
      },
    });
  });

  test('can use FnRef where it is expected to be evaluated to FnRef (not an object)', async () => {
    // GIVEN
    const source = {
      Resources: {
        MyBucket: {
          Type: 'aws-cdk-lib.aws_s3.Bucket',
        },
        TestRole: {
          Type: 'AWS::IAM::Role',
          Properties: {
            assumedBy: {
              'aws-cdk-lib.aws_iam.AccountPrincipal': { Ref: 'AWS::AccountId' },
            },
            inlinePolicies: {
              bucketAccess: {
                'aws-cdk-lib.aws_iam.PolicyStatement.fromJson': {
                  obj: {
                    Effect: 'Allow',
                    Action: ['s3:GetObject*', 's3:PutObject*'],
                    Resource: [
                      { 'CDK::GetProp': 'MyBucket.bucketArn' },
                      { Ref: 'MyBucket' },
                      {
                        'Fn::Join': [
                          '',
                          [{ 'Fn::GetAtt': 'MyBucket.Arn' }, '/*'],
                        ],
                      },
                      {
                        'Fn::Join': ['', [{ Ref: 'MyBucket' }, '/*']],
                      },
                    ],
                  },
                },
              },
            },
          },
        },
      },
    };
    const template = await Testing.template(Template.fromObject(source));

    // THEN
    expect(template).toMatchSnapshot();
  });

  test('can use FnRef as string property', async () => {
    // GIVEN
    const source = {
      Resources: {
        CfnBucket: {
          Type: 'AWS::S3::Bucket',
        },
        CdkBucket: {
          Type: 'aws-cdk-lib.aws_s3.Bucket',
        },
        Topic: {
          Type: 'aws-cdk-lib.aws_sns.Topic',
          Properties: {
            displayName: { Ref: 'CfnBucket' },
            topicName: { Ref: 'CdkBucket' },
          },
        },
      },
    };
    const template = await Testing.template(Template.fromObject(source));

    // THEN
    template.hasResourceProperties('AWS::SNS::Topic', {
      DisplayName: { Ref: 'CfnBucket' },
      TopicName: { Ref: Match.stringLikeRegexp('^CdkBucket.{8}$') },
    });
  });

  test('can use FnRef on result of Member Method Call', async () => {
    // GIVEN
    const source = {
      Resources: {
        MyLambda: {
          Type: 'aws-cdk-lib.aws_lambda.Function',
          Properties: {
            handler: 'app.hello_handler',
            runtime: 'PYTHON_3_9',
            code: {
              'aws-cdk-lib.aws_lambda.Code.fromAsset':
                'examples/lambda-handler',
            },
          },
        },
        Alias: {
          Type: 'aws-cdk-lib.aws_lambda.Alias',
          Call: [
            'MyLambda',
            {
              addAlias: 'live',
            },
          ],
        },
        Topic: {
          Type: 'aws-cdk-lib.aws_sns.Topic',
          Properties: {
            displayName: { Ref: 'Alias' },
            topicName: { 'CDK::GetProp': 'Alias.aliasName' },
          },
        },
      },
    };
    const template = await Testing.template(Template.fromObject(source));

    // THEN
    template.hasResourceProperties('AWS::SNS::Topic', {
      DisplayName: {
        Ref: Match.stringLikeRegexp('^MyLambdaAliaslive.{8}$'),
      },
      TopicName: 'live',
    });
  });

  test('can use FnGetProp on a construct with the same logical Id as a Mapping or Output', async () => {
    // GIVEN
    const template = await Testing.template(
      await Template.fromObject({
        Mappings: {
          TopicOne: {
            'ap-southeast-1': {
              Name: 'Value01',
            },
          },
        },
        Resources: {
          TopicOne: {
            Type: 'aws-cdk-lib.aws_sns.Topic',
            Properties: {
              topicName: 'one',
              fifo: true,
            },
          },
          TopicTwo: {
            Type: 'aws-cdk-lib.aws_sns.Topic',
            Properties: {
              topicName: 'two',
              fifo: { 'CDK::GetProp': 'TopicOne.fifo' },
            },
          },
        },
        Outputs: {
          TopicOne: {
            Value: {
              Ref: 'TopicOne',
            },
          },
        },
      })
    );

    // THEN
    template.hasResourceProperties('AWS::SNS::Topic', {
      TopicName: 'two.fifo', // .fifo is added automatically for fifo topics
      FifoTopic: true,
    });
  });
});
