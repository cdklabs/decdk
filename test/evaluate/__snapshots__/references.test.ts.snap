// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can use FnRef where it is expected to be evaluated to FnRef (not an object) 1`] = `
Object {
  "Resources": Object {
    "MyBucketF68F3FF0": Object {
      "DeletionPolicy": "Retain",
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "TestRole": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "assumedBy": Object {
          "aws-cdk-lib.aws_iam.AccountPrincipal": Object {
            "Ref": "AWS::AccountId",
          },
        },
        "inlinePolicies": Object {
          "bucketAccess": Object {
            "aws-cdk-lib.aws_iam.PolicyStatement.fromJson": Object {
              "obj": Object {
                "Action": Array [
                  "s3:GetObject*",
                  "s3:PutObject*",
                ],
                "Effect": "Allow",
                "Resource": Array [
                  Object {
                    "Fn::GetAtt": Array [
                      "MyBucketF68F3FF0",
                      "Arn",
                    ],
                  },
                  Object {
                    "Ref": "MyBucketF68F3FF0",
                  },
                  Object {
                    "Fn::Join": Array [
                      "",
                      Array [
                        Object {
                          "Fn::GetAtt": Array [
                            "MyBucketF68F3FF0",
                            "Arn",
                          ],
                        },
                        "/*",
                      ],
                    ],
                  },
                  Object {
                    "Fn::Join": Array [
                      "",
                      Array [
                        Object {
                          "Ref": "MyBucketF68F3FF0",
                        },
                        "/*",
                      ],
                    ],
                  },
                ],
              },
            },
          },
        },
      },
      "Type": "AWS::IAM::Role",
      "UpdateReplacePolicy": "Delete",
    },
  },
}
`;
